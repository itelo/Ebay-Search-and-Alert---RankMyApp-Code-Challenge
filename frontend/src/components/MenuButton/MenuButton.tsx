import React from "react";
import Paper from "@material-ui/core/Paper";

import ClickAwayListener from "@material-ui/core/ClickAwayListener";
import Grow from "@material-ui/core/Grow";
import Popper from "@material-ui/core/Popper";
import MenuItem from "@material-ui/core/MenuItem";
import MenuList from "@material-ui/core/MenuList";
import Button from "@material-ui/core/Button";

type CustomizedInputBaseProps = {
  message: string;
  onChange(value: "2" | "10" | "30"): void;
  data: DataT[];
};

type DataT = {
  label: string;
  value: "2" | "10" | "30";
};

const CustomizedInputBase: React.FC<CustomizedInputBaseProps> = props => {
  const [open, setOpen] = React.useState(false);
  const anchorRef = React.useRef<HTMLButtonElement>(null);

  function handleToggle() {
    setOpen(prevOpen => !prevOpen);
  }

  function handleClose(intervalDesired?: "2" | "10" | "30") {
    return (event: React.MouseEvent<EventTarget>) => {
      if (intervalDesired) {
        props.onChange(intervalDesired);
      }
      if (
        anchorRef.current &&
        anchorRef.current.contains(event.target as HTMLElement)
      ) {
        // @ts-ignore
        return;
      }

      setOpen(false);
    };
  }

  return (
    <div>
      <Button
        ref={anchorRef}
        aria-controls="menu-list-grow"
        aria-haspopup="true"
        onClick={handleToggle}
      >
        {props.message}
      </Button>
      <Popper
        open={open}
        anchorEl={anchorRef.current}
        keepMounted={false}
        transition
        disablePortal
      >
        {({ TransitionProps, placement }) => (
          <Grow
            {...TransitionProps}
            style={{
              transformOrigin:
                placement === "bottom" ? "center top" : "center bottom"
            }}
          >
            <Paper id="menu-list-grow">
              <ClickAwayListener onClickAway={handleClose()}>
                <MenuList>
                  {props.data.map(item => (
                    <MenuItem
                      key={item.value}
                      onClick={handleClose(item.value)}
                    >
                      {item.label}
                    </MenuItem>
                  ))}
                </MenuList>
              </ClickAwayListener>
            </Paper>
          </Grow>
        )}
      </Popper>
    </div>
  );
};

export default CustomizedInputBase;
