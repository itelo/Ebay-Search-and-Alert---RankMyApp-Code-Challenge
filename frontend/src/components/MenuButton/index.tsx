import MenuItem from "./MenuButton";

export default MenuItem;
