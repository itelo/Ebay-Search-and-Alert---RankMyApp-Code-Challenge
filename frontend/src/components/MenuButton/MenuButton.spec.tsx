import React from "react";
import { render, fireEvent, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import "@testing-library/dom";
import MenuButton from "./MenuButton";

afterEach(cleanup);

jest.mock(
  "popper.js",
  () =>
    class Popper {
      static placements = [
        "auto",
        "auto-end",
        "auto-start",
        "bottom",
        "bottom-end",
        "bottom-start",
        "left",
        "left-end",
        "left-start",
        "right",
        "right-end",
        "right-start",
        "top",
        "top-end",
        "top-start"
      ];

      constructor() {
        return {
          destroy: () => {},
          scheduleUpdate: () => {}
        };
      }
    }
);

describe("<MenuButton />", () => {
  it("test basic rendering when close", () => {
    const mockFn = jest.fn();
    const props = {
      message: "menu-button-message",
      onChange: mockFn,
      data: [
        {
          value: "2" as "2",
          label: "2 min"
        }
      ]
    };
    const { getByTestId, queryByTestId, asFragment } = render(
      <MenuButton {...props} />
    );

    expect(getByTestId("button-message-tag")).toHaveTextContent(
      "menu-button-message"
    );
    expect(queryByTestId("menu-item-tag-2")).toBeNull();

    expect(asFragment()).toMatchSnapshot();
  });

  it("test basic rendering when open", () => {
    const mockFn = jest.fn();
    const props = {
      message: "menu-button-message",
      onChange: mockFn,
      data: [
        {
          value: "2" as "2",
          label: "2 min"
        }
      ]
    };
    const { getByTestId, queryByTestId, asFragment, debug } = render(
      <MenuButton {...props} />
    );

    fireEvent.click(getByTestId("button-message-tag"));

    expect(getByTestId("button-message-tag")).toHaveTextContent(
      "menu-button-message"
    );

    expect(getByTestId("menu-item-tag-2")).toHaveTextContent("2 min");

    expect(asFragment()).toMatchSnapshot();
  });

  it("test click to handle change", () => {
    const mockFn = jest.fn();
    const props = {
      message: "menu-button-message",
      onChange: mockFn,
      data: [
        {
          value: "2" as "2",
          label: "2 min"
        }
      ]
    };
    const { getByTestId, queryByTestId, asFragment, debug } = render(
      <MenuButton {...props} />
    );

    fireEvent.click(getByTestId("button-message-tag"));
    fireEvent.click(getByTestId("menu-item-tag-2"));

    expect(mockFn).toHaveBeenCalled();
  });
});
