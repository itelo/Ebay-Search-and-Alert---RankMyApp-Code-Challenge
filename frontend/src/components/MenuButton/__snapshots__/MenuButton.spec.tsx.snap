// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MenuButton /> test basic rendering when close 1`] = `
<DocumentFragment>
  <div>
    <button
      aria-controls="menu-list-grow"
      aria-haspopup="true"
      class="MuiButtonBase-root MuiButton-root MuiButton-text"
      data-testid="button-message-tag"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        menu-button-message
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`<MenuButton /> test basic rendering when open 1`] = `
<DocumentFragment>
  <div>
    <button
      aria-controls="menu-list-grow"
      aria-haspopup="true"
      class="MuiButtonBase-root MuiButton-root MuiButton-text"
      data-testid="button-message-tag"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        menu-button-message
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <div
      role="tooltip"
      style="position: absolute;"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
        id="menu-list-grow"
        style="opacity: 1; transform: scale(1, 1); transform-origin: center top; transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      >
        <ul
          class="MuiList-root MuiList-padding"
          role="menu"
          tabindex="-1"
        >
          <li
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
            data-testid="menu-item-tag-2"
            role="menuitem"
            tabindex="-1"
          >
            2 min
            <span
              class="MuiTouchRipple-root"
            />
          </li>
        </ul>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
